src/odemis/driver/orsay.py

```diff
@@ -2614,3 +2641,247 @@ def terminate(self):
+    objective lens a is a constant coefficient stored in MD_CALIB (should equal 0.18e6 V/m) and d is the relative change
+    in focus distance in meter.
+    """
+
+    def __init__(self, name, role, parent, rng, metadata, **kwargs):
+        """
+        Initialise Focus. Raises AttributeError exception if there is no _fib_beam sibling.
+        :param (tuple (float, float)) rng: the range of the z axis
+        """
+        axes_def = {"z": model.Axis(unit="m", range=rng)}
+        super().__init__(name, role, parent=parent, axes=axes_def, **kwargs)
+
+        self._metadata.update(metadata)  # store the metadata passed through the config file
+
+        self.position = model.VigilantAttribute({"z": 0.0}, readonly=True, unit="m")
+        self.parent._fib_beam.objectiveVoltage.subscribe(self._updatePosition)
+
+        self.baseLensVoltage = 0.0  # V1, the objective lens voltage corresponding to a focus distance of 0
+        # Changes during runtime
+
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)
+
+    def _updatePosition(self, value=None):
+        """
+        Calculates the current focus distance as d = (value - baseLensVoltage) / MD_CALIB
+        :param (float or None) value: the current value of the objective lens voltage. If value is None, this voltage
+        will be read from the FIBBeam sibling.
+        """
+        if value is None:
+            value = self.parent._fib_beam.objectiveVoltage.value
+        new_d = (value - self.baseLensVoltage) / self._metadata[model.MD_CALIB]
+        self.position._set_value({"z": new_d}, force_write=True)
+
+    def _doMoveRel(self, shift, timeout=60):
+        """
+        Calculate the new position of the focus and pass the result to _doMoveAbs
+        :param (dict {"z": value}) shift: value contains the desired change in focus position in meter.
+        :param (int or float) timeout: the maximum number of seconds to wait until a TimeoutError is raised.
+        """
+        shift["z"] += self.position.value["z"]
+        self._doMoveAbs(shift, timeout)
+
+    def _doMoveAbs(self, pos, timeout=60):
+        """
+        Calculated the new voltage as V = baseLensVoltage + MD_CALIB * (current_position + delta).
+        Blocking until the new position is reached or it times out.
+        :param (dict {"z": value}) pos: value contains the desired new focus position in meter.
+        :param (int or float) timeout: the maximum number of seconds to wait until a TimeoutError is raised.
+        """
+        new_voltage = self.baseLensVoltage + self._metadata[model.MD_CALIB] * pos["z"]
+        self.parent._fib_beam.objectiveVoltage.value = new_voltage
+
+        start = time.time()
+        while not self.position.value["z"] == pos['z']:
+            if time.time() - start >= timeout:
+                raise TimeoutError("Timed out changing the objective lens voltage")
```

Comment:
There are quite a few problems with this loop. First, it's an active loop. It'll use the CPU at 100% while waiting. Second, if the .position is not computed exactly identical value it'll block until the timeout. Third, it'd be nice to indicate the timeout duration in the error message, the requested position, and current position.

=> Instead, subscribe to .position (before changing the objectiveVoltage), to a small listener function that sets an Event. Then, while the value is not `util.almost_equal()`, wait on the Event until the end time (computed at the beginning). You could also log every time the event is triggered but the position is concidered too far, to see if it's really blocking.